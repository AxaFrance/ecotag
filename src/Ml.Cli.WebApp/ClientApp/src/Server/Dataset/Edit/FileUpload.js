import React from 'react';
import { File, FileTable } from '@axa-fr/react-toolkit-form-input-file';
import Button from '@axa-fr/react-toolkit-button';
import cuid from 'cuid';
import {resilienceStatus} from "../../shared/Resilience";

export const FileUpload = ({fetch, setState, state}) => {

    const onChange = value => {
        const reader = new FileReader();
        reader.onloadend = () => {
            setState({ ...state, files : { ...state.files, filesLoad: value }});
        };
        reader.readAsDataURL(value.values[0].file);
    };

    const deleteFile = (idFile) => {
        const files = state.files;
        const filesLoad = files.filesLoad;
        const file = filesLoad.values.find(v => v.id === idFile);
        const index = filesLoad.values.indexOf(file);
        const newValues = [...filesLoad.values];
        if (index > -1) {
            newValues.splice(index, 1);
        }
        setState({ ...state, files : { ...state.files, filesLoad: { values : newValues }  }});
    };

    const sendFile = async() => {

        setState({...state, status : resilienceStatus.POST });
     
            let i, j, temporary, chunk = 10;
            const array = state.files.filesLoad.values;
            const promises = [];
            for (i = 0,j = array.length; i < j; i += chunk) {
                temporary = array.slice(i, i + chunk);
                const formData = new FormData();
                for (const value of temporary) {
                    formData.append('files', value.file);
                }
                const responsePromise = fetch(`datasets/${state.dataset.id}/files`, {
                    method: 'POST',
                    header: {
                        'content-type': 'multipart/form-data',
                    },
                    body: formData,
                });
                promises.push(responsePromise)
            }

        const responses = await Promise.all(promises);
            if(responses.find(response => response.status >= 500)){
                setState({...state, status : resilienceStatus.ERROR });
            } else {
                const newFilesSend = state.files.filesSend;
                state.files.filesLoad.values.map(file => {
                    const newFile = {...file, id: cuid()};
                    newFilesSend.push(newFile);
                });
                setState({...state, files: {...state.files, filesLoad: [], filesSend: newFilesSend}});
            }
    };

    return (
        <>
            <div className={`edit-dataset__file-upload-container edit-dataset__file-upload-container--${state.dataset.isLock ? 'disabled' : ''}`}>
                <h2 className="edit-dataset__file-upload-title">Upload des fichiers {state.dataset.type === "Image" ? '.jpg, .jpeg, .png, .tiff': '.txt'}</h2>
                <File
                    id='file'
                    name='datasetUploadFiles'
                    onChange={onChange}
                    multiple={true}
                    isVisible={true}
                    readOnly={false}
                    disabled={false}
                    accept={state.dataset.type === "Image" ? 'image/jpeg, image/png, image/tiff': 'text/plain'}
                    label="Parcourir"
                    icon='open'
                />
                {state.files.filesLoad.length === 0 ? '' : <FileTable errors={[]} values={state.files.filesLoad.values} onClick={deleteFile} />}
                <div>
                    <Button
                        name="sendFiles"
                        id="sendFiles"
                        onClick={sendFile}
                        disabled={state.files.filesLoad.length === 0}
                        classModifier={state.files.filesLoad.length === 0 ? 'disabled' : ''}>
                        <span className="af-btn__text">Envoyer</span>
                    </Button>
                </div>
            </div>
        </>
    );
};

export default FileUpload;
