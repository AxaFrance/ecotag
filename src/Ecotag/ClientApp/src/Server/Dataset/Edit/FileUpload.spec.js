import {render} from '@testing-library/react';
import {FileUpload} from "./FileUpload";

describe('FileUpload', () => {

    const state = {
        files: {
            filesSend: [],
            filesSendError: [],
            filesLoad: [],
            filesLoadError: [],
            paging: {itemByPages: 50, currentPages: 1}
        },
        openLockModal: false,
        dataset: {
            id: '0001',
            name: 'datasetName',
            createDate: 0,
            type: 'Image',
            isLock: false
        }
    };

    const cases = ["Image", "Text", "Eml"];
    test.each(cases)(
        "given %p should render correctly",
        (datasetType) => {
            const newState = {...state, dataset: {...state.dataset, type: datasetType}};
            const {asFragment} = render(<FileUpload state={newState} setState={() => {
            }} fetch={() => {
            }}/>);
            expect(asFragment()).toMatchSnapshot();
        }
    );

})