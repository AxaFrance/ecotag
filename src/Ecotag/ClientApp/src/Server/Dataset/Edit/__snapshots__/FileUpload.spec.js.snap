// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUpload given "Eml" should render correctly in english 1`] = `
<DocumentFragment>
  <div
    class="edit-dataset__file-upload-container edit-dataset__file-upload-container--disabled"
  >
    <h2
      class="edit-dataset__file-upload-title"
    >
      Upload files .eml
    </h2>
    <div
      class="af-form__file-input"
    >
      <div
        class="drop-box hidden-mobile"
        id="file"
        tabindex="0"
      >
        <input
          accept="message/rfc822, application/octet-stream"
          autocomplete="off"
          multiple=""
          name="datasetUploadFiles"
          style="display: none;"
          tabindex="-1"
          type="file"
        />
        <div>
          Drag/drop your files here
        </div>
      </div>
      <button
        class="af-btn af-btn--file af-btn--hasIconLeft"
        type="button"
      >
        <i
          class="glyphicon glyphicon-open"
        />
         Browse
      </button>
    </div>
    <div>
      <button
        class="btn af-btn af-btn--disabled"
        disabled=""
        id="sendFiles"
        name="sendFiles"
      >
        <span
          class="af-btn__text"
        >
          Send
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FileUpload given "Image" should render correctly in english 1`] = `
<DocumentFragment>
  <div
    class="edit-dataset__file-upload-container edit-dataset__file-upload-container--disabled"
  >
    <h2
      class="edit-dataset__file-upload-title"
    >
      Upload files .jpg, .jpeg, .png
    </h2>
    <div
      class="af-form__file-input"
    >
      <div
        class="drop-box hidden-mobile"
        id="file"
        tabindex="0"
      >
        <input
          accept="image/jpeg, image/png"
          autocomplete="off"
          multiple=""
          name="datasetUploadFiles"
          style="display: none;"
          tabindex="-1"
          type="file"
        />
        <div>
          Drag/drop your files here
        </div>
      </div>
      <button
        class="af-btn af-btn--file af-btn--hasIconLeft"
        type="button"
      >
        <i
          class="glyphicon glyphicon-open"
        />
         Browse
      </button>
    </div>
    <div>
      <button
        class="btn af-btn af-btn--disabled"
        disabled=""
        id="sendFiles"
        name="sendFiles"
      >
        <span
          class="af-btn__text"
        >
          Send
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FileUpload given "Text" should render correctly in english 1`] = `
<DocumentFragment>
  <div
    class="edit-dataset__file-upload-container edit-dataset__file-upload-container--disabled"
  >
    <h2
      class="edit-dataset__file-upload-title"
    >
      Upload files .txt
    </h2>
    <div
      class="af-form__file-input"
    >
      <div
        class="drop-box hidden-mobile"
        id="file"
        tabindex="0"
      >
        <input
          accept="text/plain"
          autocomplete="off"
          multiple=""
          name="datasetUploadFiles"
          style="display: none;"
          tabindex="-1"
          type="file"
        />
        <div>
          Drag/drop your files here
        </div>
      </div>
      <button
        class="af-btn af-btn--file af-btn--hasIconLeft"
        type="button"
      >
        <i
          class="glyphicon glyphicon-open"
        />
         Browse
      </button>
    </div>
    <div>
      <button
        class="btn af-btn af-btn--disabled"
        disabled=""
        id="sendFiles"
        name="sendFiles"
      >
        <span
          class="af-btn__text"
        >
          Send
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FileUpload should render correctly in french 1`] = `
<DocumentFragment>
  <div
    class="edit-dataset__file-upload-container edit-dataset__file-upload-container--disabled"
  >
    <h2
      class="edit-dataset__file-upload-title"
    >
      Upload des fichiers .jpg, .jpeg, .png
    </h2>
    <div
      class="af-form__file-input"
    >
      <div
        class="drop-box hidden-mobile"
        id="file"
        tabindex="0"
      >
        <input
          accept="image/jpeg, image/png"
          autocomplete="off"
          multiple=""
          name="datasetUploadFiles"
          style="display: none;"
          tabindex="-1"
          type="file"
        />
        <div>
          Glissez/d√©posez vos fichiers ici
        </div>
      </div>
      <button
        class="af-btn af-btn--file af-btn--hasIconLeft"
        type="button"
      >
        <i
          class="glyphicon glyphicon-open"
        />
         Parcourir
      </button>
    </div>
    <div>
      <button
        class="btn af-btn af-btn--disabled"
        disabled=""
        id="sendFiles"
        name="sendFiles"
      >
        <span
          class="af-btn__text"
        >
          Envoyer
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
